File: internal/models/host.go
// agency-gateway/internal/models/host.go
package models

// ComponentLimits defines container resource limits per component.
type ComponentLimits struct {
	// Per-agent components
	EnforcerMemMB             int `yaml:"enforcer_mem_mb" default:"32"`
	EnforcerCPUMilli          int `yaml:"enforcer_cpu_milli" default:"500"`
	WorkspaceMemMB            int `yaml:"workspace_mem_mb" default:"512"`
	WorkspaceBodyMemMB        int `yaml:"workspace_body_mem_mb" default:"256"`
	WorkspaceMemReservationMB int `yaml:"workspace_mem_reservation_mb" default:"256"`
	WorkspaceCPUMilli         int `yaml:"workspace_cpu_milli" default:"1000"`

	// Shared infrastructure
	EgressMemMB    int `yaml:"egress_mem_mb" default:"256"`
	CommsMemMB     int `yaml:"comms_mem_mb" default:"128"`
	KnowledgeMemMB int `yaml:"knowledge_mem_mb" default:"128"`
	IntakeMemMB    int `yaml:"intake_mem_mb" default:"128"`

	// Headroom multipliers
	EnforcerHeadroom  float64 `yaml:"enforcer_headroom" default:"4.0"`
	WorkspaceHeadroom float64 `yaml:"workspace_headroom" default:"16.0"`
	EgressHeadroom    float64 `yaml:"egress_headroom" default:"4.0"`
}

// PerAgentMB returns total memory per agent.
func (c *ComponentLimits) PerAgentMB() int {
	return c.EnforcerMemMB + c.WorkspaceBodyMemMB
}

// SharedInfraMB returns total memory for shared infrastructure.
func (c *ComponentLimits) SharedInfraMB() int {
	return c.EgressMemMB + c.CommsMemMB +
		c.KnowledgeMemMB + c.IntakeMemMB
}

// HostProfile holds detected host resources and derived capacity limits.
type HostProfile struct {
	CPUCount        int            `yaml:"cpu_count" validate:"required"`
	TotalRAMMB      int            `yaml:"total_ram_mb" validate:"required"`
	AvailableDiskGB int            `yaml:"available_disk_gb" validate:"required"`
	OSReserveMB     int            `yaml:"os_reserve_mb" default:"512"`
	ComponentLimits ComponentLimits `yaml:"component_limits"`
	MaxAgents       int            `yaml:"max_agents"`
	ImageBaselines  map[string]int `yaml:"image_baselines"`
}

// Validate computes max_agents from host resources.
func (h *HostProfile) Validate() error {
	h.ComputeMaxAgents()
	return nil
}

// ComputeMaxAgents calculates max agents from available RAM.
func (h *HostProfile) ComputeMaxAgents() {
	available := h.TotalRAMMB - h.ComponentLimits.SharedInfraMB() - h.OSReserveMB
	perAgent := h.ComponentLimits.PerAgentMB()
	if perAgent > 0 {
		h.MaxAgents = available / perAgent
		if h.MaxAgents < 1 {
			h.MaxAgents = 1
		}
	} else {
		h.MaxAgents = 1
	}
}

// ApplyBaselines recomputes limits from image baselines.
func (h *HostProfile) ApplyBaselines() {
	mapping := map[string]struct {
		field    *int
		headroom float64
	}{
		"enforcer":  {&h.ComponentLimits.EnforcerMemMB, h.ComponentLimits.EnforcerHeadroom},
		"workspace": {&h.ComponentLimits.WorkspaceMemMB, h.ComponentLimits.WorkspaceHeadroom},
		"egress":         {&h.ComponentLimits.EgressMemMB, h.ComponentLimits.EgressHeadroom},
	}
	for image, m := range mapping {
		if baseline, ok := h.ImageBaselines[image]; ok && baseline > 0 {
			*m.field = int(float64(baseline) * m.headroom)
		}
	}
	h.ComputeMaxAgents()
}

File: images/models/host.py
"""Host profile model -- machine resources and container limits."""

from __future__ import annotations

from pydantic import BaseModel, model_validator


class ComponentLimits(BaseModel):
    """Container resource limits per component, in MB and millicpus."""

    # Per-agent components
    enforcer_mem_mb: int = 32
    enforcer_cpu_milli: int = 500
    gateway_mem_mb: int = 64
    gateway_cpu_milli: int = 500
    workspace_mem_mb: int = 512
    workspace_body_mem_mb: int = 256
    workspace_mem_reservation_mb: int = 256
    workspace_cpu_milli: int = 1000

    # Shared infrastructure
    egress_mem_mb: int = 256
    comms_mem_mb: int = 128
    knowledge_mem_mb: int = 128
    intake_mem_mb: int = 128
    vault_mem_mb: int = 128
    vault_sidecar_mem_mb: int = 64

    # Headroom multipliers (applied to image baselines when available)
    enforcer_headroom: float = 4.0
    gateway_headroom: float = 4.0
    workspace_headroom: float = 16.0
    egress_headroom: float = 4.0

    @property
    def per_agent_mb(self) -> int:
        """Total memory per agent (enforcer + gateway + workspace body)."""
        return self.enforcer_mem_mb + self.gateway_mem_mb + self.workspace_body_mem_mb

    @property
    def shared_infra_mb(self) -> int:
        """Total memory for shared infrastructure (egress + comms + swarm vault)."""
        return (
            self.egress_mem_mb + self.comms_mem_mb
            + self.knowledge_mem_mb + self.intake_mem_mb
            + self.vault_mem_mb + self.vault_sidecar_mem_mb
        )


class HostProfile(BaseModel):
    """Detected host resources and derived capacity limits."""

    cpu_count: int
    total_ram_mb: int
    available_disk_gb: int
    os_reserve_mb: int = 512
    component_limits: ComponentLimits = ComponentLimits()
    max_agents: int = 1
    image_baselines: dict[str, int] = {}

    @model_validator(mode="after")
    def _compute_max_agents(self) -> "HostProfile":
        limits = self.component_limits
        available = self.total_ram_mb - limits.shared_infra_mb - self.os_reserve_mb
        if limits.per_agent_mb > 0:
            computed = max(1, available // limits.per_agent_mb)
        else:
            computed = 1
        self.max_agents = computed
        return self

    def apply_baselines(self) -> None:
        """Recompute component limits from image baselines if available."""
        limits = self.component_limits
        mapping = {
            "enforcer": ("enforcer_mem_mb", limits.enforcer_headroom),
            "workspace": ("workspace_mem_mb", limits.workspace_headroom),
            "egress": ("egress_mem_mb", limits.egress_headroom),
        }
        for image, (field_name, headroom) in mapping.items():
            baseline = self.image_baselines.get(image)
            if baseline and baseline > 0:
                setattr(limits, field_name, int(baseline * headroom))
        self._compute_max_agents()
